File: event_organizer.php
# event_organizer.php
#!/usr/bin/env php
<?php

define('DATA_FILE', 'events.json');

class Task {
    public $id;
    public $description;
    public $category;
    public $completed;

    function __construct($description, $category = '') {
        $this->id = substr(bin2hex(random_bytes(4)), 0, 8);
        $this->description = $description;
        $this->category = $category;
        $this->completed = false;
    }

    function toArray() {
        return ['id' => $this->id, 'description' => $this->description,
                'category' => $this->category, 'completed' => $this->completed];
    }

    static function fromArray($data) {
        $task = new self($data['description'], $data['category']);
        $task->id = $data['id'];
        $task->completed = $data['completed'] ?? false;
        return $task;
    }
}

class Event {
    public $id;
    public $name;
    public $date;
    public $description;
    public $tasks = [];

    function __construct($name, $date = '', $description = '') {
        $this->id = substr(bin2hex(random_bytes(4)), 0, 8);
        $this->name = $name;
        $this->date = $date;
        $this->description = $description;
    }

    function toArray() {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'date' => $this->date,
            'description' => $this->description,
            'tasks' => array_map(function($t) { return $t->toArray(); }, $this->tasks)
        ];
    }

    static function fromArray($data) {
        $ev = new self($data['name'], $data['date'], $data['description']);
        $ev->id = $data['id'];
        $ev->tasks = array_map(function($t) { return Task::fromArray($t); }, $data['tasks'] ?? []);
        return $ev;
    }
}

class Organizer {
    private $events = [];

    function __construct() {
        $this->load();
    }

    function load() {
        if (file_exists(DATA_FILE)) {
            $data = json_decode(file_get_contents(DATA_FILE), true);
            $this->events = array_map(function($e) { return Event::fromArray($e); }, $data);
        }
    }

    function save() {
        file_put_contents(DATA_FILE, json_encode(array_map(function($e) { return $e->toArray(); }, $this->events), JSON_PRETTY_PRINT));
    }

    function getEvent($id) {
        foreach ($this->events as $e) {
            if ($e->id == $id) return $e;
        }
        return null;
    }

    function create($name, $date = '', $description = '') {
        $ev = new Event($name, $date, $description);
        $this->events[] = $ev;
        $this->save();
        echo "✅ Event created: {$ev->name} (ID: {$ev->id})\n";
    }

    function addTask($eventId, $desc, $category = '') {
        $ev = $this->getEvent($eventId);
        if (!$ev) {
            echo "Event $eventId not found.\n";
            return;
        }
        $task = new Task($desc, $category);
        $ev->tasks[] = $task;
        $this->save();
        echo "✅ Task added: {$task->description} (ID: {$task->id})\n";
    }

    function list() {
        if (empty($this->events)) {
            echo "No events.\n";
            return;
        }
        echo "\n📋 Events:\n";
        foreach ($this->events as $e) {
            $total = count($e->tasks);
            $done = 0;
            foreach ($e->tasks as $t) if ($t->completed) $done++;
            $pct = $total > 0 ? round($done / $total * 100) : 0;
            echo "  {$e->name} (ID: {$e->id}) – {$done}/{$total} tasks done ({$pct}%)\n";
        }
    }

    function show($eventId) {
        $ev = $this->getEvent($eventId);
        if (!$ev) {
            echo "Event $eventId not found.\n";
            return;
        }
        $total = count($ev->tasks);
        $done = 0;
        foreach ($ev->tasks as $t) if ($t->completed) $done++;
        $pct = $total > 0 ? round($done / $total * 100) : 0;
        echo "\n📋 Event: {$ev->name}\n";
        if ($ev->date) echo "Date: {$ev->date}\n";
        if ($ev->description) echo "Description: {$ev->description}\n";
        echo "Progress: $done/$total tasks done ($pct%)\n";
        if ($total == 0) {
            echo "No tasks yet.\n";
            return;
        }
        echo "\nTasks:\n";
        foreach ($ev->tasks as $i => $task) {
            $status = $task->completed ? '[X]' : '[ ]';
            $cat = $task->category ? " ({$task->category})" : '';
            echo "  " . ($i+1) . ". $status {$task->description}$cat (ID: {$task->id})\n";
        }
    }

    function markDone($eventId, $taskId) {
        $ev = $this->getEvent($eventId);
        if (!$ev) {
            echo "Event $eventId not found.\n";
            return;
        }
        foreach ($ev->tasks as $task) {
            if ($task->id == $taskId) {
                if ($task->completed) {
                    echo "Task already completed.\n";
                } else {
                    $task->completed = true;
                    $this->save();
                    echo "✅ Task marked as done: {$task->description}\n";
                }
                return;
            }
        }
        echo "Task $taskId not found in this event.\n";
    }

    function delete($eventId) {
        foreach ($this->events as $i => $e) {
            if ($e->id == $eventId) {
                $name = $e->name;
                array_splice($this->events, $i, 1);
                $this->save();
                echo "✅ Event deleted: $name\n";
                return;
            }
        }
        echo "Event $eventId not found.\n";
    }
}

if ($argc < 2) {
    die("Usage: php event_organizer.php <command> [options]\n");
}
$app = new Organizer();
$cmd = $argv[1];

switch ($cmd) {
    case 'create':
        if ($argc < 3) die("create <name> [--date DATE] [--desc TEXT]\n");
        $name = $argv[2];
        $date = '';
        $desc = '';
        for ($i=3; $i<$argc; $i++) {
            if ($argv[$i] == '--date' && isset($argv[$i+1])) { $date = $argv[++$i]; }
            if ($argv[$i] == '--desc' && isset($argv[$i+1])) { $desc = $argv[++$i]; }
        }
        $app->create($name, $date, $desc);
        break;

    case 'add-task':
        if ($argc < 4) die("add-task <event_id> <description> [--category CAT]\n");
        $eventId = $argv[2];
        $desc = $argv[3];
        $category = '';
        for ($i=4; $i<$argc; $i++) {
            if ($argv[$i] == '--category' && isset($argv[$i+1])) { $category = $argv[++$i]; }
        }
        $app->addTask($eventId, $desc, $category);
        break;

    case 'list':
        $app->list();
        break;

    case 'show':
        if ($argc < 3) die("show <event_id>\n");
        $app->show($argv[2]);
        break;

    case 'done':
        if ($argc < 4) die("done <event_id> <task_id>\n");
        $app->markDone($argv[2], $argv[3]);
        break;

    case 'delete':
        if ($argc < 3) die("delete <event_id>\n");
        $app->delete($argv[2]);
        break;

    default:
        echo "Unknown command.\n";
}
?>

File: event_organizer.cpp
// event_organizer.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <random>
#include <nlohmann/json.hpp>

using namespace std;
using json = nlohmann::json;

class Task {
public:
    string id;
    string description;
    string category;
    bool completed;

    Task() : completed(false) {}
    Task(const string& desc, const string& cat = "") : description(desc), category(cat), completed(false) {
        id = generateId();
    }

    string generateId() {
        const char* hex = "0123456789abcdef";
        string id;
        random_device rd;
        mt19937 gen(rd());
        uniform_int_distribution<> dis(0, 15);
        for (int i=0; i<8; i++) id += hex[dis(gen)];
        return id;
    }

    json toJson() const {
        return {{"id", id}, {"description", description}, {"category", category}, {"completed", completed}};
    }

    static Task fromJson(const json& j) {
        Task t;
        t.id = j["id"];
        t.description = j["description"];
        t.category = j.value("category", "");
        t.completed = j.value("completed", false);
        return t;
    }
};

class Event {
public:
    string id;
    string name;
    string date;
    string description;
    vector<Task> tasks;

    Event() {}
    Event(const string& n, const string& d = "", const string& desc = "") : name(n), date(d), description(desc) {
        id = generateId();
    }

    string generateId() {
        const char* hex = "0123456789abcdef";
        string id;
        random_device rd;
        mt19937 gen(rd());
        uniform_int_distribution<> dis(0, 15);
        for (int i=0; i<8; i++) id += hex[dis(gen)];
        return id;
    }

    json toJson() const {
        json tasksJson = json::array();
        for (const auto& t : tasks) tasksJson.push_back(t.toJson());
        return {{"id", id}, {"name", name}, {"date", date}, {"description", description}, {"tasks", tasksJson}};
    }

    static Event fromJson(const json& j) {
        Event e;
        e.id = j["id"];
        e.name = j["name"];
        e.date = j.value("date", "");
        e.description = j.value("description", "");
        for (const auto& t : j["tasks"]) {
            e.tasks.push_back(Task::fromJson(t));
        }
        return e;
    }
};

class Organizer {
private:
    vector<Event> events;
    string dataFile = "events.json";

    void load() {
        ifstream f(dataFile);
        if (!f.is_open()) return;
        json j;
        f >> j;
        for (auto& item : j) {
            events.push_back(Event::fromJson(item));
        }
    }

    void save() {
        json j = json::array();
        for (auto& e : events) j.push_back(e.toJson());
        ofstream f(dataFile);
        f << setw(2) << j << endl;
    }

    Event* getEvent(const string& id) {
        for (auto& e : events) {
            if (e.id == id) return &e;
        }
        return nullptr;
    }

public:
    Organizer() { load(); }

    void create(const string& name, const string& date, const string& desc) {
        Event e(name, date, desc);
        events.push_back(e);
        save();
        cout << "✅ Event created: " << e.name << " (ID: " << e.id << ")\n";
    }

    void addTask(const string& eventId, const string& desc, const string& category) {
        Event* e = getEvent(eventId);
        if (!e) {
            cout << "Event " << eventId << " not found.\n";
            return;
        }
        Task t(desc, category);
        e->tasks.push_back(t);
        save();
        cout << "✅ Task added: " << t.description << " (ID: " << t.id << ")\n";
    }

    void list() {
        if (events.empty()) {
            cout << "No events.\n";
            return;
        }
        cout << "\n📋 Events:\n";
        for (auto& e : events) {
            int total = e.tasks.size();
            int done = 0;
            for (auto& t : e.tasks) if (t.completed) done++;
            int pct = total > 0 ? (int)(done * 100.0 / total) : 0;
            cout << "  " << e.name << " (ID: " << e.id << ") – " << done << "/" << total << " tasks done (" << pct << "%)\n";
        }
    }

    void show(const string& eventId) {
        Event* e = getEvent(eventId);
        if (!e) {
            cout << "Event " << eventId << " not found.\n";
            return;
        }
        int total = e->tasks.size();
        int done = 0;
        for (auto& t : e->tasks) if (t.completed) done++;
        int pct = total > 0 ? (int)(done * 100.0 / total) : 0;
        cout << "\n📋 Event: " << e->name << "\n";
        if (!e->date.empty()) cout << "Date: " << e->date << "\n";
        if (!e->description.empty()) cout << "Description: " << e->description << "\n";
        cout << "Progress: " << done << "/" << total << " tasks done (" << pct << "%)\n";
        if (total == 0) {
            cout << "No tasks yet.\n";
            return;
        }
        cout << "\nTasks:\n";
        int i = 1;
        for (auto& t : e->tasks) {
            string status = t.completed ? "[X]" : "[ ]";
            string cat = t.category.empty() ? "" : " (" + t.category + ")";
            cout << "  " << i++ << ". " << status << " " << t.description << cat << " (ID: " << t.id << ")\n";
        }
    }

    void markDone(const string& eventId, const string& taskId) {
        Event* e = getEvent(eventId);
        if (!e) {
            cout << "Event " << eventId << " not found.\n";
            return;
        }
        for (auto& t : e->tasks) {
            if (t.id == taskId) {
                if (t.completed) {
                    cout << "Task already completed.\n";
                } else {
                    t.completed = true;
                    save();
                    cout << "✅ Task marked as done: " << t.description << "\n";
                }
                return;
            }
        }
        cout << "Task " << taskId << " not found in this event.\n";
    }

    void del(const string& eventId) {
        for (auto it = events.begin(); it != events.end(); ++it) {
            if (it->id == eventId) {
                string name = it->name;
                events.erase(it);
                save();
                cout << "✅ Event deleted: " << name << "\n";
                return;
            }
        }
        cout << "Event " << eventId << " not found.\n";
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Usage: event_organizer <command> [options]\n";
        return 1;
    }
    Organizer app;
    string cmd = argv[1];

    if (cmd == "create") {
        if (argc < 3) { cerr << "create <name> [--date DATE] [--desc TEXT]\n"; return 1; }
        string name = argv[2];
        string date, desc;
        for (int i=3; i<argc; i++) {
            if (string(argv[i]) == "--date" && i+1 < argc) date = argv[++i];
            if (string(argv[i]) == "--desc" && i+1 < argc) desc = argv[++i];
        }
        app.create(name, date, desc);
    } else if (cmd == "add-task") {
        if (argc < 4) { cerr << "add-task <event_id> <description> [--category CAT]\n"; return 1; }
        string eventId = argv[2];
        string desc = argv[3];
        string category;
        for (int i=4; i<argc; i++) {
            if (string(argv[i]) == "--category" && i+1 < argc) category = argv[++i];
        }
        app.addTask(eventId, desc, category);
    } else if (cmd == "list") {
        app.list();
    } else if (cmd == "show") {
        if (argc < 3) { cerr << "show <event_id>\n"; return 1; }
        app.show(argv[2]);
    } else if (cmd == "done") {
        if (argc < 4) { cerr << "done <event_id> <task_id>\n"; return 1; }
        app.markDone(argv[2], argv[3]);
    } else if (cmd == "delete") {
        if (argc < 3) { cerr << "delete <event_id>\n"; return 1; }
        app.del(argv[2]);
    } else {
        cerr << "Unknown command.\n";
        return 1;
    }
    return 0;
}
